pilot/pkg/proxy/envoy/v2/ads.go

```diff
@@ -645,6 +652,8 @@ func (con *XdsConnection) send(res *xdsapi.DiscoveryResponse) error {
 		err := con.stream.Send(res)
 		done <- err
 		if res.Nonce != "" {
+			con.mutex.Lock()
```

Comment:
Looking at this now, is the `con.stream.Send` actually safe? gRPC streams cannot be written to concurrently, so if this function is racing to set the field in the con struct, it's likely also racing to write to the stream.

Either way, rather than defering the unlock I prefer just wrapping the switch in the lock/unlock:
```go
if res.Nonce != "' {
    con.mutex.Lock()
    switch res.TypeUrl {
        // ...
    }
    con.mutex.Unlock()
}

(total nit, not blocking)

Reply:
@ZackButcher I appreciate any comment even if just nits.
The mutex is protection against r/w of the struct fields as detected by the racetest tool. I don't mind moving the mutex lock to before calling the Send but I tend to avoid keeping locks when doing some lengthy I/O operations and would actually expect the Send itself to be protected by its implementation.

Reply:
No, that's my point: the gRPC stream implementation is not safe for concurrent rights. See the [documentation](https://godoc.org/google.golang.org/grpc#Stream). Specifically: "But it is not safe to call SendMsg on the same stream in different goroutines."